File: api/routers/evals.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, field_validator
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_db
from deps import get_current_user
from ml.evaluation import build_eval_payload

router = APIRouter(prefix="/api/v1/evals", tags=["evals"])

VALID_BANDS = {"blue", "green", "yellow", "red"}
VALID_VERDICTS = {"hit", "miss"}

class EvalPostRequest(BaseModel):
    round_id: int
    predicted_band: str
    actual_band: str
    actual_multiplier: float
    verdict: str
    evaluated_at: datetime | None = None

    @field_validator("predicted_band", "actual_band")
    @classmethod
    def validate_band(cls, v: str) -> str:
        if v not in VALID_BANDS:
            raise ValueError(f"band must be one of {VALID_BANDS}")
        return v

    @field_validator("verdict")
    @classmethod
    def validate_verdict(cls, v: str) -> str:
        if v not in VALID_VERDICTS:
            raise ValueError(f"verdict must be one of {VALID_VERDICTS}")
        return v

    @field_validator("actual_multiplier")
    @classmethod
    def validate_multiplier(cls, v: float) -> float:
        if not (1.01 <= v <= 501.00):
            raise ValueError("actual_multiplier must be between 1.01 and 501.00")
        return round(v, 2)


@router.post("", status_code=201)
async def post_eval(
    body: EvalPostRequest,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    evaluated_at = body.evaluated_at or datetime.now(timezone.utc)
    await db.execute(
        text(
            "INSERT INTO prediction_evals "
            "(round_id, predicted_band, actual_band, actual_multiplier, verdict, evaluated_at) "
            "VALUES (:round_id, :predicted_band, :actual_band, :actual_multiplier, :verdict, :evaluated_at) "
            "ON CONFLICT (round_id) DO NOTHING"
        ),
        {
            "round_id": body.round_id,
            "predicted_band": body.predicted_band,
            "actual_band": body.actual_band,
            "actual_multiplier": body.actual_multiplier,
            "verdict": body.verdict,
            "evaluated_at": evaluated_at,
        },
    )
    await db.commit()
    return {"ok": True}


@router.get("/stats")
async def get_eval_stats(
    limit: int = Query(default=500, ge=1, le=5000),
    db: AsyncSession = Depends(get_db),
):
    rows_result = await db.execute(
        text(
            "SELECT predicted_band, actual_band, verdict, COUNT(*) AS cnt "
            "FROM prediction_evals "
            "GROUP BY predicted_band, actual_band, verdict "
            "ORDER BY predicted_band, verdict"
        )
    )
    rows = rows_result.fetchall()

    # 全体の的中率
    total_row_result = await db.execute(
        text("SELECT COUNT(*) AS total, SUM(CASE WHEN verdict='hit' THEN 1 ELSE 0 END) AS hits FROM prediction_evals")
    )
    t = total_row_result.fetchone()
    total = t.total or 0
    hits = t.hits or 0

    # 直近 N 件
    recent_rows_result = await db.execute(
        text(
            "SELECT round_id, predicted_band, actual_band, actual_multiplier, verdict, evaluated_at "
            "FROM prediction_evals ORDER BY evaluated_at DESC LIMIT :lim"
        ),
        {"lim": limit},
    )
    recent_rows = recent_rows_result.fetchall()

    # レスポンス整形中はDB接続を保持しない
    await db.rollback()

    stats = [
        {"predicted_band": r.predicted_band, "actual_band": r.actual_band,
         "verdict": r.verdict, "count": r.cnt}
        for r in rows
    ]
    recent = [
        {
            "round_id": r.round_id,
            "predicted_band": r.predicted_band,
            "actual_band": r.actual_band,
            "actual_multiplier": float(r.actual_multiplier),
            "verdict": r.verdict,
            "evaluated_at": r.evaluated_at.isoformat() if hasattr(r.evaluated_at, "isoformat") else str(r.evaluated_at),
        }
        for r in recent_rows
    ]

    return {
        "total": total,
        "hits": hits,
        "hit_rate": round(hits / total, 4) if total > 0 else None,
        "by_band": stats,
        "recent": recent,
    }


@router.delete("", status_code=200)
async def delete_evals(
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    """prediction_evals のみをリセットする（rounds は保持）。"""
    count_row = await db.execute(text("SELECT COUNT(*) FROM prediction_evals"))
    deleted = count_row.scalar()
    await db.execute(text("TRUNCATE prediction_evals"))
    await db.commit()
    return {"deleted": deleted}


@router.post("/rebuild", status_code=200)
async def rebuild_evals(
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    """現行モデルで prediction_evals を全件再計算する。"""
    rounds_rows = await db.execute(
        text("SELECT id, multiplier FROM rounds ORDER BY recorded_at ASC, id ASC")
    )
    rounds = [{"id": int(r.id), "multiplier": float(r.multiplier)} for r in rounds_rows]

    await db.execute(text("TRUNCATE prediction_evals"))

    history: list[float] = []
    payloads: list[dict] = []
    for row in rounds:
        payload = build_eval_payload(
            round_id=row["id"],
            actual_multiplier=row["multiplier"],
            history=history,
        )
        if payload is not None:
            payloads.append(payload)

        history.append(row["multiplier"])
        history = history[-WINDOW:]

    if payloads:
        await db.execute(
            text(
                "INSERT INTO prediction_evals "
                "(round_id, predicted_band, actual_band, actual_multiplier, verdict) "
                "VALUES (:round_id, :predicted_band, :actual_band, :actual_multiplier, :verdict)"
            ),
            payloads,
        )

    await db.commit()
    return {
        "rounds": len(rounds),
        "rebuilt": len(payloads),
        "skipped_head": min(len(rounds), WINDOW),
    }

File: frontend/app/lib/evals.ts
export type PredictedBand = "blue" | "green" | "yellow" | "red";

export type EvalStatsRow = {
  predicted_band: PredictedBand;
  actual_band: PredictedBand;
  verdict: "hit" | "miss";
  count: number;
};

export type EvalRecentRow = {
  round_id: number;
  predicted_band: PredictedBand;
  actual_band: PredictedBand;
  actual_multiplier: number;
  verdict: "hit" | "miss";
  evaluated_at: string;
};

export type EvalStatsResponse = {
  by_band: EvalStatsRow[];
  recent: EvalRecentRow[];
};

export type RoundEval = {
  predicted_band: PredictedBand;
  actual_band: PredictedBand;
  predicted_line?: number;
  actual: number;
  verdict: "hit" | "miss";
  emoji: string;
  label: string;
  evaluated_at: string;
};

export type RoundEvalArchive = Record<string, RoundEval>;

export type PredictionSummaryRow = {
  band: PredictedBand;
  predictedCount: number;
  hitCount: number;
  overCount: number | null;
  hitRate: number | null;
};

const SUMMARY_BANDS: PredictedBand[] = ["red", "yellow", "green", "blue"];

export function bandLabel(band: PredictedBand): string {
  if (band === "blue") return "1x台";
  if (band === "green") return "2x以上";
  if (band === "yellow") return "5x以上";
  return "10x以上";
}

export function bandLabelJa(band: PredictedBand): string {
  if (band === "blue") return "青";
  if (band === "green") return "緑";
  if (band === "yellow") return "黄";
  return "赤";
}

function isOverResult(predictedBand: PredictedBand, actualBand: PredictedBand): boolean {
  if (predictedBand === "green") return actualBand === "yellow" || actualBand === "red";
  if (predictedBand === "yellow") return actualBand === "red";
  return false;
}

export function buildPredictionSummary(rows: EvalStatsRow[]): PredictionSummaryRow[] {
  return SUMMARY_BANDS.map((band) => {
    const relevant = rows.filter((row) => row.predicted_band === band);
    const predictedCount = relevant.reduce((sum, row) => sum + row.count, 0);
    const hitCount = relevant
      .filter((row) => row.verdict === "hit")
      .reduce((sum, row) => sum + row.count, 0);
    const overCount = band === "blue" || band === "red"
      ? null
      : relevant
          .filter((row) => isOverResult(row.predicted_band, row.actual_band))
          .reduce((sum, row) => sum + row.count, 0);

    return {
      band,
      predictedCount,
      hitCount,
      overCount,
      hitRate: predictedCount > 0 ? (hitCount / predictedCount) * 100 : null,
    };
  });
}

type BuildEvalArchiveOptions = {
  missEmoji?: string;
  labelSeparator?: string;
};

export function buildEvalArchive(
  rows: EvalRecentRow[],
  options?: BuildEvalArchiveOptions,
): RoundEvalArchive {
  const missEmoji = options?.missEmoji ?? "●";
  const labelSeparator = options?.labelSeparator ?? " ";
  const archive: RoundEvalArchive = {};
  for (const row of rows) {
    archive[String(row.round_id)] = {
      predicted_band: row.predicted_band,
      actual_band: row.actual_band,
      actual: row.actual_multiplier,
      verdict: row.verdict,
      emoji: row.verdict === "hit" ? "✅" : missEmoji,
      label: `予測:${bandLabel(row.predicted_band)}${labelSeparator}実績:${bandLabel(row.actual_band)}`,
      evaluated_at: row.evaluated_at,
    };
  }
  return archive;
}
